Packages/com.chisel.graph/Chisel/ChiselGraphInstance.cs

```diff
@@ -0,0 +1,131 @@
+using Chisel.Core;
+using System;
+using System.Collections.Generic;
+using System.Reflection;
+using Unity.Collections;
+using Unity.Jobs;
+using UnityEngine;
+
+namespace Chisel.Nodes
+{
+    [ExecuteInEditMode]
+    public class ChiselGraphInstance : MonoBehaviour
+    {
+        public ChiselGraph graph;
+        public MeshFilter meshFilter;
+        public MeshRenderer meshRenderer;
+        public bool IsDirty { get; set; }
+
+        public List<GraphProperty> properties;
+        public Dictionary<string, GraphProperty> overriddenProperties;
+
+        CSGTree tree;
+        List<Mesh> meshes;
+
+        void Start()
+        {
+            meshes = new List<Mesh>();
+            UpdateCSG();
+        }
+
+        void Update()
+        {
+            UpdateCSG();
+        }
+
+        void OnValidate()
+        {
+            graph.instance = this;
+
+            if (properties == null || graph.properties.Count != properties.Count)
+                InitProperties();
+
+            for (int i = 0; i < graph.properties.Count; i++)
+                if (graph.properties[i].Name != properties[i].Name)
+                    InitProperties();
+        }
+
+        void InitProperties()
+        {
+            properties = new List<GraphProperty>();
+            foreach (var property in graph.properties)
+                properties.Add(Clone(property));
+            UpdateProperties();
+        }
+
+        GraphProperty Clone(GraphProperty source)
+        {
+            var newProperty = Activator.CreateInstance(source.GetType()) as GraphProperty;
+            var fields = source.GetType().GetFields(BindingFlags.Public | BindingFlags.Instance);
+
+            foreach (var info in fields)
+                info.SetValue(newProperty, info.GetValue(source));
+
+            return newProperty;
+        }
+
+        public void UpdateProperties()
+        {
+            if (overriddenProperties == null)
+                overriddenProperties = new Dictionary<string, GraphProperty>();
+            overriddenProperties.Clear();
+
+            foreach (var property in properties)
+                if (property.overrideValue)
+                    overriddenProperties[property.Name] = property;
+        }
+
+        public void UpdateCSG()
+        {
+            if (!IsDirty) return;
+            IsDirty = false;
+
+            CSGManager.Clear();
+
+            graph.instance = this;
+
+            if (!tree.Valid)
+                tree = CSGTree.Create(GetInstanceID());
+            else
+                tree.Clear();
```

Comment:
At some point we need a new API where instead of you creating a tree here, it would create a branch, and this branch would be inserted in another tree. I'm working on redesigning this part of Chisel for this reason

Reply:
Do you mean like how I wrap a subgraph in a branch? Or like a global tree that contain everything?

Reply:
Each tree creates it's own separate meshes and they do not interact with each other, so in order to have a graph that can be used together with other brushes & generators, they need to be part of the same tree.
Right now the mechanism for that is less than ideal, so I'm redesigning it & after that the graph could just create a branch that can then be hooked into the tree that encompasses all generators and brushes

Reply:
Right, you are talking about making the graph instance work with chisel components, I wonder where should the integration goes, I would prefer keeping the node graph depend on the chisel core only. the use case I have is parametric modeling of a in game item, instead of level editing. 
Making graph depends on component, or component depends on graph doesn't seem quite right. Maybe a third package "Graph Component" that depends on both packages?

Reply:
Well no, the components will work on top of chisel core as well. Just having one single graph to make everything is a bit limiting, so we need to have a solution where multiple graphs can each create pieces of a tree